contracts/forwarder/Eip712Forwarder.sol

```diff
@@ -0,0 +1,145 @@
+    }
+
+    function verify(ForwardRequest memory req,
+        bytes32 domainSeparator, bytes32 requestTypeHash, bytes memory suffixData, bytes memory sig) public override view {
+
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+    }
+
+    function verifyAndCall(ForwardRequest memory req,
+        bytes32 domainSeparator, bytes32 requestTypeHash, bytes memory suffixData, bytes memory sig) public override {
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+        _updateNonce(req);
+
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success,) = req.to.call{gas: req.gas}(abi.encodePacked(req.data, req.from));
+        if (!success) {
+            // solhint-disable-next-line no-inline-assembly
+            assembly {// This assembly ensure the revert contains the exact string data
+                let returnDataSize := returndatasize()
+                returndatacopy(0, 0, returnDataSize)
+                revert(0, returnDataSize)
+            }
+        }
+    }
+
+
+    function _verifyNonce(ForwardRequest memory req) internal view {
+        require(nonces[req.from] == req.nonce, "nonce mismatch");
+    }
+
+    function _updateNonce(ForwardRequest memory req) internal {
+        nonces[req.from]++;
+    }
+
+    //register a requestTypeHash
+    // the given requestType must have GENERIC_PARAMS as the beginning of the type string,
+    // otherwise it can't be registered.
+
+    /**
+     * Register a new Request typehash.
+     * @param typeName - the name of the request type.
+     * @param extraParams - params to add to the request type, after initial "_ForwardRequest request" param
+     * @param subTypes - subtypes used by the extraParamsg
+     * @param subTypes2 - more subtypes, if sorted after _ForwardRequest (e.g. if type starts with lowercase)
+     */
+    function registerRequestType(string calldata typeName, string calldata extraParams, string calldata subTypes, string calldata subTypes2) external {
+
+        require(bytes(typeName).length > 0, "invalid typeName");
+        bytes memory types = abi.encodePacked(subTypes, GENERIC_TYPE, subTypes2);
+        string memory comma = bytes(extraParams).length > 0 ? "," : "";
+
+        bytes memory requestType = abi.encodePacked(
+            typeName, "(",
+            GENERIC_PARAMS, comma, extraParams, ")",
+            types
+        );
+        bytes32 requestTypehash = keccak256(bytes(requestType));
+        uint len = bytes(subTypes).length;
+        //sanity: avoid redefining our type, e.g.: subType="_ForwardRequest(whatever)_z"
+        require(len == 0 || bytes(subTypes)[len - 1] == ")", "invalid subType");
+        //sanity: parameters should not end parameters block
+        for (uint i = 0; i < bytes(extraParams).length; i++) {
+            require(bytes(extraParams)[i] != ")", "invalid extraParams");
+        }
```

Comment:
I am not sure this is equivalent, here we are talking about a malicious JS code trying to trick user into signing something, not an executable code. So, the severity is lover.

Reply:
the only reason for using a structure, is make sure the user sees what we meant him to see,
e.g. if someone can swap the "from" and "to" fields, the user might think the tx means different thing.
That's why its important there is no way to abuse the structure.
Since there is no "trusted authority", we had to check it on chain.
we can simplify by far the check if instead of struct we have plain fields.